subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps_utils.c

```diff
@@ -0,0 +1,521 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <pm_config.h>
+#include <stdlib.h>
+
+#include <net/nrf_cloud_pgps.h>
+#include <settings/settings.h>
+#include <net/download_client.h>
+#include <date_time.h>
+
+#include "nrf_cloud_transport.h"
+#include "nrf_cloud_pgps_schema_v1.h"
+#include "nrf_cloud_pgps_utils.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(nrf_cloud_pgps, CONFIG_NRF_CLOUD_GPS_LOG_LEVEL);
+
+#define SETTINGS_NAME				"nrf_cloud_pgps"
+#define SETTINGS_KEY_PGPS_HEADER		"pgps_header"
+#define SETTINGS_FULL_PGPS_HEADER		SETTINGS_NAME "/" SETTINGS_KEY_PGPS_HEADER
+#define SETTINGS_KEY_LOCATION			"location"
+#define SETTINGS_FULL_LOCATION			SETTINGS_NAME "/" SETTINGS_KEY_LOCATION
+#define SETTINGS_KEY_LEAP_SEC			"g2u_leap_sec"
+#define SETTINGS_FULL_LEAP_SEC			SETTINGS_NAME "/" SETTINGS_KEY_LEAP_SEC
+
+struct block_pool {
+	int first_free;
+	uint32_t last_alloc;
+	bool block_used[NUM_PREDICTIONS];
+};
+
+static struct block_pool pool;
+static uint8_t *block_pool_base;
+static int num_blocks;
+
+/* todo: get the correct values from somewhere */
+static int gps_leap_seconds = GPS_TO_UTC_LEAP_SECONDS;
```

Comment:
I don't know where we could get the correct number of leap seconds during runtime. GNSS gets this information from satellite broadcast, but GNSS only indicates whether it has decoded the number of leap seconds from the broadcast and the current UTC time. The number of leap seconds is not available and even if it would be, the library would most likely still need help from the application to get this information.

Another problem is that to decode the leap second from the broadcast, GNSS would most likely have to run for a long time. Downloading all pages which contain also this information takes 12.5 minutes, if I remember correctly. In the usual use case where GNSS runs periodically for short periods of time, GNSS will most likely never get this information.

One possibility which comes to my mind would be to operate in GPS time when possible, but that would most likely require the application to inject the current GPS time into the library after each fix. And before the first fix the best estimate of the GPS time would have to be used anyway.

Reply:
If A-GPS is enabled as well, my modifications to `nrf_cloud_agps.c` in `agps_send_to_modem()` grabs this value from the `NRF_CLOUD_AGPS_UTC_PARAMETERS` in the `delta_tls` field.

Ideally, we would add this information to the PGPS file header, so a P-GPS-only build would get this information as well.  It's too late to add this for NCS 1.6.0 but could be added after that.

I'm guessing there may be multiple places this information could be culled from automatically on the cloud side.

Reply:
Including this in the PGPS data from the cloud sounds like a good idea. I think this is good enough for now, but a better solution is needed in the future.